web/src/main/java/org/cbioportal/web/util/StudyViewFilterUtil.java

```diff
@@ -153,4 +156,117 @@ public Integer getFilteredCountByDataEquality(List<ClinicalDataFilter> attribute
             return clinicalDataCountItem;
         }).collect(Collectors.toList());
     }
+    
+    public boolean shouldSkipFilterForClinicalDataBins(StudyViewFilter filter) {
+        // if everything other than study ids and sample identifiers is null,
+        // we can skip the filter for data bin calculation
+        return (
+            filter != null &&
+            filter.getClinicalDataFilters() == null &&
+            filter.getGeneFilters() == null &&
+            filter.getSampleTreatmentFilters() == null &&
+            filter.getPatientTreatmentFilters() == null &&
+            filter.getGenomicProfiles() == null &&
+            filter.getGenomicDataFilters() == null &&
+            filter.getGenericAssayDataFilters() == null &&
+            filter.getCaseLists() == null &&
+            filter.getCustomDataFilters() == null
+        );
+    }
+    
+    public List<ClinicalData> filterClinicalData(
+        List<ClinicalData> unfilteredClinicalDataForSamples,
+        List<ClinicalData> unfilteredClinicalDataForPatients,
+        List<ClinicalData> unfilteredClinicalDataForConflictingPatientAttributes,
+        List<String> studyIds,
+        List<String> sampleIds,
+        List<String> studyIdsOfPatients,
+        List<String> patientIds,
+        List<String> sampleAttributeIds,
+        List<String> patientAttributeIds,
+        List<String> conflictingPatientAttributes
+    ) {
+        List<ClinicalData> combinedResult = new ArrayList<>();
+        
+        // create look up for faster filtering
+        Map<String, String> sampleIdToStudyId = mapCaseToStudy(sampleIds, studyIds);
+        Map<String, String> patientIdToStudyId = mapCaseToStudy(patientIds, studyIdsOfPatients);
+        Map<String, Boolean> sampleAttributeIdLookup = listToMap(sampleAttributeIds);
+        Map<String, Boolean> patientAttributeIdLookup = listToMap(patientAttributeIds);
+        Map<String, Boolean> conflictingPatientAttributeIdLookup = listToMap(conflictingPatientAttributes);
```

Comment:
Ideally we should not create indexes for every single API call. If we find a way to cache indexes like we cache the DB query responses, we can further improve the performance of filtering.

Reply:
Profiling shows that the time required to create indexes is insignificant compared to the time required for the rest of the operations. Let's keep it this way for now.

Reply:
Looks like these Maps are required depending on attributes. may be they can be moved inside `if` block